lang/cs/Org.Apache.REEF.Network/Group/Task/Impl/GroupCommNetworkObserver.cs

```diff
@@ -16,84 +16,81 @@
-        /// WritableCommunicationGroupNetworkObserver.
+        /// Registers a <see cref="TaskMessageObserver"/> for a given <see cref="taskSourceId"/>.
+        /// If if the <see cref="TaskMessageObserver"/> has already been initialized, it will return
+        /// the existing one.
         /// </summary>
-        /// <param name="nsMessage"></param>
-        public void OnNext(NsMessage<GeneralGroupCommunicationMessage> nsMessage)
+        public TaskMessageObserver RegisterAndGetForTask(string taskSourceId)
         {
-            if (nsMessage == null)
-            {
-                throw new ArgumentNullException("nsMessage");
-            }
-
-            try
-            {
-                GeneralGroupCommunicationMessage gcm = nsMessage.Data.First();
-                _commGroupHandlers[gcm.GroupName].OnNext(gcm);
-            }
-            catch (InvalidOperationException)
-            {
-                LOGGER.Log(Level.Error, "Group Communication Network Handler received message with no data");
-                throw;
-            }
-            catch (KeyNotFoundException)
-            {
-                LOGGER.Log(Level.Error, "Group Communication Network Handler received message for nonexistant group");
-                throw;
-            }
+            // Add a TaskMessage observer for each upstream/downstream source.
+            return _taskMessageObservers.GetOrAdd(taskSourceId, new TaskMessageObserver(_networkService.Get()));
         }
 
-        /// <summary>
-        /// Registers the network handler for the given CommunicationGroup.
-        /// When messages are sent to the specified group name, the given handler
-        /// will be invoked with that message.
-        /// </summary>
-        /// <param name="groupName">The group name for the network handler</param>
-        /// <param name="commGroupHandler">The network handler to invoke when
-        /// messages are sent to the given group.</param>
-        public void Register(string groupName, IObserver<GeneralGroupCommunicationMessage> commGroupHandler)
+        public void OnNext(IRemoteMessage<NsMessage<GeneralGroupCommunicationMessage>> remoteMessage)
         {
-            if (string.IsNullOrEmpty(groupName))
+            try
             {
-                throw new ArgumentNullException("groupName");
+                var nsMessage = remoteMessage.Message;
+                var gcm = nsMessage.Data.First();
+                var gcMessageTaskSource = gcm.Source;
+                TaskMessageObserver observer;
+                if (!_taskMessageObservers.TryGetValue(gcMessageTaskSource, out observer))
+                {
+                    throw new KeyNotFoundException("Unable to find registered NodeMessageObserver for source Task " +
+                                                   gcMessageTaskSource + ".");
+                }
+
+                _registeredNodes.GetOrAdd(gcMessageTaskSource,
```

Comment:
You cannot check if an item exists in `ConcurrentBag`. There is no `ConcurrentSet`.